runtime/compiler/control/JITaaSCompilationThread.cpp

```diff
@@ -3260,175 +3289,196 @@ TR::CompilationInfoPerThreadRemote::processEntry(TR_MethodToBeCompiled &entry, J
-
-      stream->setClientId(clientId);
-      setSeqNo(seqNo); // memorize the sequence number of this request
-
-      bool sessionDataWasEmpty = false;
-      ClientSessionData *clientSession = NULL;
-      {
-      // Get a pointer to this client's session data
-      // Obtain monitor RAII style because creating a new hastable entry may throw bad_alloc
-      OMR::CriticalSection compilationMonitorLock(compInfo->getCompilationMonitor());
-      compInfo->getClientSessionHT()->purgeOldDataIfNeeded(); // Try to purge old data
-      if (!(clientSession = compInfo->getClientSessionHT()->findOrCreateClientSession(clientId, seqNo, &sessionDataWasEmpty)))
-         throw std::bad_alloc();
-
-      setClientData(clientSession); // Cache the session data into CompilationInfoPerThreadRemote object
-      if (TR::Options::getVerboseOption(TR_VerboseJITaaS))
-         TR_VerboseLog::writeLineLocked(TR_Vlog_JITaaS, "compThreadID=%d %s clientSessionData=%p for clientUID=%llu seqNo=%u",
-            getCompThreadId(), sessionDataWasEmpty ? "created" : "found", clientSession, (unsigned long long)clientId, seqNo);
-      } // end critical section
-
-     
-      // We must process unloaded classes lists in the same order they were generated at the client
-      // Use a sequencing scheme to re-order compilation requests
-      //
-      clientSession->getSequencingMonitor()->enter();
-      clientSession->updateMaxReceivedSeqNo(seqNo);
-      if (seqNo > clientSession->getExpectedSeqNo()) // out of order messages
+      if (!ramMethod)
          {
-         // park this request until the missing ones arrive
          if (TR::Options::getVerboseOption(TR_VerboseJITaaS))
-            TR_VerboseLog::writeLineLocked(TR_Vlog_JITaaS, "Out-of-sequence msg detected for clientUID=%llu seqNo=%u > expectedSeqNo=%u. Parking compThreadID=%d (entry=%p)",
-            (unsigned long long)clientId, seqNo, clientSession->getExpectedSeqNo(), getCompThreadId(), &entry);
-
-         waitForMyTurn(clientSession, entry);
+            TR_VerboseLog::writeLineLocked(TR_Vlog_JITaaS,"Message to delete client (%lu) received", clientId);
+         compInfo->acquireCompMonitor(compThread); //need to acquire compilation monitor for both deleting the client data and the setting the thread state to COMPTHREAD_SIGNAL_SUSPEND.
+         bool result = compInfo->getClientSessionHT()->deleteClientSession(clientId, true);
+         if (!result)
+            {
+            if (TR::Options::getVerboseOption(TR_VerboseJITaaS))
+               TR_VerboseLog::writeLineLocked(TR_Vlog_JITaaS,"client (%lu) not deleted", clientId);
+            }
+         else
+            {
+            if (TR::Options::getVerboseOption(TR_VerboseJITaaS))
+               TR_VerboseLog::writeLineLocked(TR_Vlog_JITaaS,"client (%lu) deleted", clientId);
+            }
+         noCompilation = true;   // This is a null message for termination so no compilation needed.
```

Comment:
Suggestion: maybe a simpler solution is to set `abortCompilation = true` and let the code take that path.

Reply:
Maybe it's not working only in this context. I have seen compilations that take this path and everything is well. Anyway, if there is something wrong with it we need to fix it.

Reply:
The way the code stands now we are leaking compilation entries because `compInfo->recycleCompilationEntry(&entry);` is not called.